File: app/src/lib/parche.js
// Parches: solo lo que cambió entre dos versiones de los datos. Mismo algoritmo que
// receptor/sincro/src/parche.rs (los dos lados se comprueban con tests/vectores/parche.json).
//
// Un parche es una lista de operaciones sobre una ruta. Cada paso de la ruta es una clave de
// objeto ("proyectos") o un elemento de una lista por su id ({ "id": "nota_1" }):
//   { r: [...], v: valor }      poner (reemplaza o agrega)
//   { r: [...], x: 1 }          quitar
//   { r: [...], orden: [ids] }  reordenar una lista con ids
import { igual } from './sincro.js'

const esObjeto = v => v !== null && typeof v === 'object' && !Array.isArray(v)
function conIds(v) {
  if (!Array.isArray(v)) return false
  const vistos = new Set()
  for (const x of v) {
    if (!esObjeto(x) || typeof x.id !== 'string' || vistos.has(x.id)) return false
    vistos.add(x.id)
  }
  return true
}

/** Operaciones que convierten `a` en `b`. */
export function diferencias(a, b, ruta = [], ops = []) {
  if (igual(a, b)) return ops
  if (esObjeto(a) && esObjeto(b)) {
    for (const k of Object.keys(b)) {
      if (b[k] === undefined) continue
      if (!(k in a) || a[k] === undefined) ops.push({ r: [...ruta, k], v: b[k] })
      else diferencias(a[k], b[k], [...ruta, k], ops)
    }
    for (const k of Object.keys(a)) if (a[k] !== undefined && (!(k in b) || b[k] === undefined)) ops.push({ r: [...ruta, k], x: 1 })
    return ops
  }
  if (conIds(a) && conIds(b)) {
    const ma = new Map(a.map(x => [x.id, x])), mb = new Set(b.map(x => x.id))
    for (const x of a) if (!mb.has(x.id)) ops.push({ r: [...ruta, { id: x.id }], x: 1 })
    for (const x of b) {
      if (!ma.has(x.id)) ops.push({ r: [...ruta, { id: x.id }], v: x })
      else diferencias(ma.get(x.id), x, [...ruta, { id: x.id }], ops)
    }
    // Tras quitar y agregar al final, ¿queda en el orden de `b`?
    const queda = [...a.filter(x => mb.has(x.id)).map(x => x.id), ...b.filter(x => !ma.has(x.id)).map(x => x.id)]
    if (queda.some((id, i) => id !== b[i].id)) ops.push({ r: ruta, orden: b.map(x => x.id) })
    return ops
  }
  ops.push({ r: ruta, v: b })
  return ops
}

const copia = v => (v === undefined ? v : JSON.parse(JSON.stringify(v)))

function hijo(contenedor, paso) {
  if (typeof paso === 'string') return esObjeto(contenedor) ? contenedor[paso] : undefined
  return Array.isArray(contenedor) ? contenedor.find(x => x?.id === paso.id) : undefined
}

/** Aplica un parche a una copia de `a` y la devuelve. */
export function aplicar(a, ops) {
  let raiz = copia(a)
  for (const op of ops) {
    if (!op.r.length) {
      if ('v' in op) raiz = copia(op.v)
      continue
    }
    let c = raiz
    for (const paso of op.r.slice(0, -1)) {
      c = hijo(c, paso)
      if (c === undefined) throw new Error('El parche no corresponde a esta versión de los datos')
    }
    const ultimo = op.r.at(-1)
    if ('orden' in op) {
      const lista = hijo(c, ultimo)
      if (!Array.isArray(lista)) throw new Error('El parche no corresponde a esta versión de los datos')
      const pos = new Map(op.orden.map((id, i) => [id, i]))
      const ordenada = [...lista].sort((x, y) => (pos.get(x.id) ?? Infinity) - (pos.get(y.id) ?? Infinity))
      lista.splice(0, lista.length, ...ordenada)
    } else if (typeof ultimo === 'string') {
      if (!esObjeto(c)) throw new Error('El parche no corresponde a esta versión de los datos')
      if (op.x) delete c[ultimo]
      else c[ultimo] = copia(op.v)
    } else {
      if (!Array.isArray(c)) throw new Error('El parche no corresponde a esta versión de los datos')
      const i = c.findIndex(x => x?.id === ultimo.id)
      if (op.x) { if (i >= 0) c.splice(i, 1) }
      else if (i >= 0) c[i] = copia(op.v)
      else c.push(copia(op.v))
    }
  }
  return raiz
}

/** JSON con las claves ordenadas: el mismo texto en JS y en Rust para los mismos datos. */
export function canonico(v) {
  if (Array.isArray(v)) return '[' + v.map(x => (x === undefined ? 'null' : canonico(x))).join(',') + ']'
  if (esObjeto(v)) {
    const ks = Object.keys(v).filter(k => v[k] !== undefined).sort()
    return '{' + ks.map(k => JSON.stringify(k) + ':' + canonico(v[k])).join(',') + '}'
  }
  return JSON.stringify(v)
}

/** Huella (sha-256, 32 hex) de los datos: ambos lados la comparan para saber si tienen la misma base. */
export async function huella(datos) {
  const v = { proyectos: datos?.proyectos || [], fuentes: datos?.fuentes || [], citas: datos?.citas || [] }
  const h = new Uint8Array(await crypto.subtle.digest('SHA-256', new TextEncoder().encode(canonico(v))))
  return [...h.subarray(0, 16)].map(b => b.toString(16).padStart(2, '0')).join('')
}

/** Tamaño aproximado (bytes) de un parche, para mostrar cuánto viajó. */
export const pesoDe = v => new TextEncoder().encode(JSON.stringify(v)).length

File: receptor/sincro/src/parche.rs
//! Parches: solo lo que cambió entre dos versiones de los datos. Mismo algoritmo que
//! `app/src/lib/parche.js` (se comprueban con `app/tests/vectores/parche.json`).
//! Operaciones: `{r, v}` poner · `{r, x: 1}` quitar · `{r, orden: [ids]}` reordenar. Cada paso de
//! la ruta `r` es una clave de objeto o `{"id": …}` (elemento de una lista por su id).
use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};
use std::collections::{HashMap, HashSet};

fn id_de(v: &Value) -> Option<&str> {
    v.as_object()?.get("id")?.as_str()
}

fn con_ids(v: &Value) -> bool {
    let Some(l) = v.as_array() else { return false };
    let mut vistos = HashSet::new();
    l.iter().all(|x| id_de(x).is_some_and(|id| vistos.insert(id)))
}

fn con(ruta: &[Value], paso: Value) -> Vec<Value> {
    let mut r = ruta.to_vec();
    r.push(paso);
    r
}

/// Operaciones que convierten `a` en `b`.
pub fn diferencias(a: &Value, b: &Value) -> Vec<Value> {
    let mut ops = vec![];
    dif(a, b, &[], &mut ops);
    ops
}

fn dif(a: &Value, b: &Value, ruta: &[Value], ops: &mut Vec<Value>) {
    if a == b {
        return;
    }
    if let (Value::Object(ma), Value::Object(mb)) = (a, b) {
        for (k, vb) in mb {
            match ma.get(k) {
                None => ops.push(json!({"r": con(ruta, json!(k)), "v": vb})),
                Some(va) => dif(va, vb, &con(ruta, json!(k)), ops),
            }
        }
        for k in ma.keys() {
            if !mb.contains_key(k) {
                ops.push(json!({"r": con(ruta, json!(k)), "x": 1}));
            }
        }
        return;
    }
    if con_ids(a) && con_ids(b) {
        let (la, lb) = (a.as_array().unwrap(), b.as_array().unwrap());
        let ma: HashMap<&str, &Value> = la.iter().map(|x| (id_de(x).unwrap(), x)).collect();
        let en_b: HashSet<&str> = lb.iter().map(|x| id_de(x).unwrap()).collect();
        for x in la {
            let id = id_de(x).unwrap();
            if !en_b.contains(id) {
                ops.push(json!({"r": con(ruta, json!({"id": id})), "x": 1}));
            }
        }
        for x in lb {
            let id = id_de(x).unwrap();
            match ma.get(id) {
                None => ops.push(json!({"r": con(ruta, json!({"id": id})), "v": x})),
                Some(va) => dif(va, x, &con(ruta, json!({"id": id})), ops),
            }
        }
        let queda: Vec<&str> = la.iter().map(|x| id_de(x).unwrap()).filter(|id| en_b.contains(id))
            .chain(lb.iter().map(|x| id_de(x).unwrap()).filter(|id| !ma.contains_key(id)))
            .collect();
        let orden: Vec<&str> = lb.iter().map(|x| id_de(x).unwrap()).collect();
        if queda != orden {
            ops.push(json!({"r": ruta, "orden": orden}));
        }
        return;
    }
    ops.push(json!({"r": ruta, "v": b}));
}

#[derive(Debug, PartialEq, Eq)]
pub struct ParcheAjeno;

fn hijo<'a>(c: &'a mut Value, paso: &Value) -> Option<&'a mut Value> {
    match paso {
        Value::String(k) => c.as_object_mut()?.get_mut(k),
        _ => {
            let id = paso.get("id")?.as_str()?;
            c.as_array_mut()?.iter_mut().find(|x| id_de(x) == Some(id))
        }
    }
}

/// Aplica un parche (lo modifica en su lugar). Error si no corresponde a estos datos.
pub fn aplicar(raiz: &mut Value, ops: &[Value]) -> Result<(), ParcheAjeno> {
    for op in ops {
        let ruta = op.get("r").and_then(|r| r.as_array()).ok_or(ParcheAjeno)?;
        let Some((ultimo, antes)) = ruta.split_last() else {
            if let Some(v) = op.get("v") {
                *raiz = v.clone();
            }
            continue;
        };
        let mut c = &mut *raiz;
        for paso in antes {
            c = hijo(c, paso).ok_or(ParcheAjeno)?;
        }
        if let Some(orden) = op.get("orden").and_then(|o| o.as_array()) {
            let lista = hijo(c, ultimo).and_then(|l| l.as_array_mut()).ok_or(ParcheAjeno)?;
            let pos: HashMap<&str, usize> = orden.iter().enumerate().filter_map(|(i, id)| Some((id.as_str()?, i))).collect();
            lista.sort_by_key(|x| id_de(x).and_then(|id| pos.get(id).copied()).unwrap_or(usize::MAX));
        } else if let Value::String(k) = ultimo {
            let m = c.as_object_mut().ok_or(ParcheAjeno)?;
            if op.get("x").is_some() {
                m.shift_remove(k);
            } else {
                m.insert(k.clone(), op.get("v").cloned().ok_or(ParcheAjeno)?);
            }
        } else {
            let id = ultimo.get("id").and_then(|i| i.as_str()).ok_or(ParcheAjeno)?;
            let l = c.as_array_mut().ok_or(ParcheAjeno)?;
            let i = l.iter().position(|x| id_de(x) == Some(id));
            if op.get("x").is_some() {
                if let Some(i) = i {
                    l.remove(i);
                }
            } else {
                let v = op.get("v").cloned().ok_or(ParcheAjeno)?;
                match i {
                    Some(i) => l[i] = v,
                    None => l.push(v),
                }
            }
        }
    }
    Ok(())
}

/// JSON con las claves ordenadas (el mismo texto que `canonico` en JS).
pub fn canonico(v: &Value) -> String {
    match v {
        Value::Array(l) => format!("[{}]", l.iter().map(canonico).collect::<Vec<_>>().join(",")),
        Value::Object(m) => {
            let mut ks: Vec<&String> = m.keys().collect();
            ks.sort_by(|a, b| a.encode_utf16().cmp(b.encode_utf16()));
            let partes: Vec<String> = ks.iter().map(|k| format!("{}:{}", Value::String((*k).clone()), canonico(&m[*k]))).collect();
            format!("{{{}}}", partes.join(","))
        }
        _ => v.to_string(),
    }
}

/// Huella de `{proyectos, fuentes, citas}` (32 hex), como `huella` en JS.
pub fn huella(datos: &Value) -> String {
    let mut m = Map::new();
    for c in crate::carpeta::COLECCIONES {
        m.insert(c.into(), datos.get(c).cloned().unwrap_or_else(|| json!([])));
    }
    let h = Sha256::digest(canonico(&Value::Object(m)).as_bytes());
    h[..16].iter().map(|b| format!("{b:02x}")).collect()
}
